### ISSUE TITLE
Wake word training GUI with voice recording and state-of-the-art TTS augmentation

### ISSUE THREAD

**Issue by krystophny** [krystophny/tabura#651]
## Summary

Build a self-contained wake word training interface inside Tabura that lets the user record their own voice saying the hotword, augment recordings with state-of-the-art voice-cloning TTS models, and train/test/deploy a custom OpenWakeWord model — all from the browser UI.

This is **Stage 3** of the wake word journey (see #649 for the full phasing):
- **Stage 1** (#649): Automatic Piper-based pre-training — fully automatic, no manual work, good enough for testing
- **Stage 2** (manual): User records voice samples via CLI, retrains — requires manual effort per iteration
- **Stage 3** (this issue): Tabura GUI replaces manual iteration — user provides initial voice samples once, then voice-cloning AI generates thousands of augmented samples automatically

Depends on #649 (sloppy hotword wiring) and #650 (pre-roll buffer integration).

## Motivation

After Stage 1, the Piper-based model works but is not tuned to the user's actual voice and accent. Stage 2 (manual CLI recording + retraining) works but requires tedious iteration. This issue eliminates that friction:

- Record voice samples directly in Tabura's UI (mic is already wired)
- Voice-cloning models (Qwen3-TTS, GPT-SoVITS) learn the user's voice from those samples
- The cloned voice then generates thousands of accent-accurate "sloppy" variants automatically — no more manual recording sessions
- Train, test in-browser, one-click deploy
- Failed detections can be recorded as new training data and fed back into the next round — the voice-cloning AI handles the augmentation, not the user

## TTS Model Strategy

Piper was chosen for real-time TTS in production. For wake word training, latency does not matter — quality and accent fidelity do. Use a hybrid approach:

### Tier 1: Voice cloning (your voice, high fidelity)

| Model | Why | Clone from | License |
|---|---|---|---|
| **Qwen3-TTS** (1.7B) | Text-describable accents ("continental European English, male, short vowels"), 3s reference | 3s audio | Apache 2.0 |
| **GPT-SoVITS** | Best fine-tuning — 1 minute of your voice is enough to fine-tune a speaker model | 1 min fine-tune | MIT |
| **Chatterbox** (Resemble AI) | Beats ElevenLabs in blind evals, paralinguistic emotion tags | 5s audio | MIT |

### Tier 2: Speaker diversity (many voices, bulk generation)

| Model | Why | License |
|---|---|---|
| **Piper** (multi-speaker VITS) | Fast CPU-only bulk generation, 100+ speakers, proven in openWakeWord | MIT |
| **Kokoro-82M** | Tiny, 96x real-time, built-in voice tensor library | Apache 2.0 |
| **Parler-TTS** | Text-described voices ("an elderly woman whispering") | Apache 2.0 |

### Pipeline

1. **Record** 20-30 samples of yourself saying "sloppy" (varied tone, speed, distance)
2. **Record** 2-3 minutes of arbitrary English speech (for voice cloning reference)
3. **Clone** your voice with Qwen3-TTS → generate 500-1000 "sloppy" samples with prosodic variation
4. **Fine-tune** GPT-SoVITS on your 1-minute recording → generate 500 more
5. **Diversify** with Piper multi-speaker + Kokoro → 1000-2000 samples across 100+ voices
6. **Augment** all samples with room impulse responses, background noise, speed/pitch perturbation (openWakeWord pipeline)
7. **Train** DNN classifier
8. **Test** in-browser with live mic
9. **Deploy** to production

### Integration into existing UI

The training GUI lives at `/hotword-train` as a standalone page (same pattern as `/capture` and `/static/hotword-test.html`). Entry points:

1. **Command center**: Add a "Train Wake Word" command to `COMMAND_CENTER_COMMANDS` in `app-command-center.ts`. Opens `/hotword-train` in the current tab (or new tab if a live session is active).

2. **Voice/hotword status**: When the hotword status indicator shows "not available" (model missing), it becomes tappable and links to `/hotword-train` with a tooltip "Set up wake word".

3. **First-run nudge**: If `GET /api/hotword/status` reports `ready: false` (no model file), the main app shows a brief non-blocking banner: "Wake word not configured — tap to set up". Tapping opens `/hotword-train`.

4. **Direct URL**: `/hotword-train` is always accessible directly for power users.

The training page is intentionally a separate page, not an in-app panel, because:
- Training is a focused, multi-step workflow that doesn't belong in the zen canvas
- It needs its own mic access independent of any live session
- The user shouldn't accidentally leave mid-training by tapping the canvas
- It mirrors the `/capture` pattern which is already a standalone page

### Page layout

Single-page vertical flow with collapsible sections. No sidebar, no edge panels — just the training workflow top to bottom. Each section has a header with status badge (not started / in progress / complete).

### Sections

**1. Voice Recording**
- Record button (reuse existing mic/audio primitives from `app-voice-audio.ts`)
- Record "sloppy" samples — shows waveform, duration, playback
- Record reference speech (longer, for voice cloning)
- List of recordings with playback, delete, download
- Target: visual counter showing "12/30 hotword samples recorded" etc.
- Audio stored as 16 kHz mono WAV in data dir

**2. TTS Generation**
- Select TTS models to use (checkboxes: Qwen3-TTS, GPT-SoVITS, Piper, Kokoro)
- Configure: number of samples per model, prosody parameters
- For Qwen3-TTS: text description field ("male, continental European accent, casual")
- Start generation → progress bar (SSE from backend)
- Preview generated samples (random sampling with playback)
- Stats: "Generated 2,847 positive samples across 4 models"

**3. Training**
- Configure: layer size, steps, threshold target, augmentation rounds
- Start training → progress bar with step count and loss curve
- Training runs the 13-step openWakeWord pipeline
- Show estimated time remaining

**4. Testing**
- Embed the existing hotword-test functionality (threshold slider, live detection)
- Load the freshly trained model for immediate testing
- Record test attempts with pass/fail marking
- "This should have triggered but didn't" → adds to next training round

**5. Deployment**
- One-click deploy: copies trained model to vendor dir, reloads hotword engine
- Rollback: keeps previous model as backup
- Shows current production model info (file size, training date, threshold)

## User-facing processing flow and progress

The user interacts with the training GUI as a single guided workflow. Here is exactly what they see at each step:

### Step 1: Record voice samples

User opens `/hotword-train`. The page shows a large record button and a counter ("0/20 hotword samples").

- User clicks record, says "sloppy", clicks stop. The recording appears in a list with waveform, duration, and playback button.
- After ~20 hotword samples, a second section asks for 2-3 minutes of general speech (for voice cloning reference). A timer shows elapsed recording time.
- Visual: green checkmark appears next to each section when enough samples are recorded.
- All recordings are uploaded immediately to the backend and stored in `{data-dir}/hotword-train/recordings/`.

### Step 2: TTS generation

User clicks "Generate training data". The page shows:

- A progress card per enabled TTS model:
  - **Qwen3-TTS**: "Cloning your voice... 0/500 samples" → progress bar fills → "234/500 samples generated"
  - **GPT-SoVITS**: "Fine-tuning on your voice (one-time)... 3/5 minutes" → "Generating... 89/500 samples"
  - **Piper**: "Generating diverse speakers... 412/1000 samples"
- Overall counter: "Generated 1,847 / 2,000 total positive samples"
- SSE stream from backend pushes live updates every ~2 seconds
- Estimated time remaining per model
- User can cancel individual models without losing already-generated samples
- If a TTS model is not installed, the card shows "Not installed — click to set up" with one-click download

### Step 3: Training

User clicks "Train model". The page shows:

- Pipeline stage indicator (augment → extract features → train DNN → export)
- During augmentation: "Augmenting... round 2/5, 3,421 samples processed"
- During training: live loss curve chart, "Step 23,450 / 50,000 — loss: 0.018"
- Estimated time remaining
- SSE stream pushes updates every ~5 seconds
- The user can leave the page and come back — backend training continues, progress resumes from SSE

### Step 4: Test

Training completes → the page automatically loads the new model into the browser's hotword engine via `initHotword({ force: true })` (no page reload needed).

- The test section activates with a live mic feed and threshold slider
- User says "sloppy" in various ways — each detection shows a green flash with confidence score
- A "This should have triggered" button records the failed attempt and adds it to a retry queue for the next training round
- A "This was a false trigger" button records false positives for negative training data
- Summary: "Detected 18/20 test attempts (90%), 2 false positives"

### Step 5: Deploy

User clicks "Deploy to production":

- Backend copies the trained `sloppy.onnx` to `internal/web/static/vendor/openwakeword/`
- Backend sends a WebSocket message to all connected clients
- All connected browsers call `initHotword({ force: true })` — the new model is live immediately, no page refresh or restart needed
- The deploy card shows: "Deployed: sloppy.onnx (142 KB, trained 2026-03-21 14:32)"
- Previous model is kept as `sloppy.onnx.bak` for one-click rollback
- If the user has the main Tabura app open in another tab, the hotword engine there also picks up the new model via the WebSocket reload signal

### Progress persistence

- If the user closes the training page mid-generation or mid-training, the backend keeps running
- Returning to `/hotword-train` reconnects the SSE stream and shows current progress
- If training completed while the page was closed, the page shows "Training complete — ready to test" with the results

### Notifications in main app

- When a new model is deployed, the main Tabura app (if open) shows a brief status overlay: "Wake word model updated"
- The `GET /api/hotword/status` endpoint includes the model timestamp and training state so any client can detect changes

## Backend APIs

All under `/api/hotword/train/` prefix, auth required.

```
POST   /api/hotword/train/recordings          Upload a voice recording (multipart WAV)
GET    /api/hotword/train/recordings          List recordings with metadata
DELETE /api/hotword/train/recordings/{id}     Delete a recording
POST   /api/hotword/train/generate            Start TTS generation (async)
GET    /api/hotword/train/generate/status     SSE stream of generation progress
POST   /api/hotword/train/start               Start model training (async)
GET    /api/hotword/train/status              SSE stream of training progress
POST   /api/hotword/train/deploy              Deploy trained model to production
GET    /api/hotword/train/models              List available trained models
POST   /api/hotword/train/test/sample         Upload a test recording for evaluation
```

### New package: `internal/hotwordtrain/`

Keeps the training orchestration out of `internal/web/`. The web layer provides HTTP handlers that delegate to this package.

Responsibilities:
- Manage recording storage (WAV files in `{data-dir}/hotword-train/recordings/`)
- Orchestrate TTS generation (shell out to Python scripts per model)
- Run openWakeWord training pipeline (`scripts/train-hotword.sh`)
- Track progress via callback/channel
- Model deployment (copy ONNX to vendor dir)

### TTS model integration

Each TTS model needs a generation script. Create `scripts/hotword-generate-{model}.sh` wrappers:

- `scripts/hotword-generate-qwen3tts.sh` — clones Qwen3-TTS repo, creates venv, generates samples from reference audio
- `scripts/hotword-generate-gptsovits.sh` — clones GPT-SoVITS, fine-tunes on reference audio, generates samples
- `scripts/hotword-generate-piper.sh` — existing Piper integration via openWakeWord
- `scripts/hotword-generate-kokoro.sh` — Kokoro-82M for fast diverse generation

Each script takes: reference audio path, output dir, number of samples, config params.
Each script outputs: WAV files in output dir + manifest JSON.

### Dependencies

- Python 3.12 with venv
- CUDA GPU recommended (CPU fallback acceptable — training is a batch job)
- ~15 GB disk for all TTS model weights
- Models are downloaded on first use, not vendored

### Files

- `internal/web/static/hotword-train.html` — page shell
- `internal/web/static/hotword-train.ts` — recording, generation, training, testing UI
- `internal/web/static/hotword-train.css` — styles

### Audio recording

Reuse patterns from `app-voice-audio.ts`:
- `getUserMedia({ audio: true })`
- `AudioContext` + `ScriptProcessor` for raw sample access
- `float32ToWav()` for WAV encoding
- Upload via `fetch()` to recording endpoint

### Progress tracking

Use Server-Sent Events (SSE) for long-running operations:
- Generation progress: "Qwen3-TTS: 234/500 samples..."
- Training progress: "Step 12,450/50,000 — loss: 0.023"
- Each SSE message includes: stage, progress %, message, ETA

### Go tests
- Recording upload/list/delete CRUD
- Training status tracking
- Model deployment (file copy + status update)

### Playwright tests
- `tests/playwright/hotword-train.spec.ts`
  - Recording flow: start → record → stop → upload → appears in list
  - Delete recording
  - Generate button triggers backend (mock)
  - Training progress SSE rendering
  - Deploy button copies model
  - Test section loads model and runs detection

## Acceptance criteria

- [ ] `/hotword-train` page accessible with auth
- [ ] Voice recording works in browser and uploads WAV
- [ ] Recordings listed with playback and delete
- [ ] TTS generation triggers backend scripts for each enabled model
- [ ] Progress streaming works via SSE
- [ ] Training runs full openWakeWord pipeline
- [ ] Trained model can be tested in-browser immediately
- [ ] One-click deployment to production vendor dir
- [ ] Failed test cases can be added to next training round
- [ ] All tests pass
